Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -426,26 +440,42 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
 
     open override func startTunnel(options: [String: NSObject]?, completionHandler: @escaping (Error?) -> Void) {
         connectionStatus = .connecting
-        tunnelHealth.isHavingConnectivityIssues = false
-        controllerErrorStore.lastErrorMessage = nil
+
+        os_log("Will load options\n%{public}@", log: .networkProtection, String(describing: options))
+        let startupOptions = StartupOptions(options: options ?? [:], log: .networkProtection)
+
+        resetIssueStateOnTunnelStart(startupOptions)
+
+        let startTime = DispatchTime.now()
 
         let internalCompletionHandler = { [weak self] (error: Error?) in
+            guard let self else {
+                completionHandler(error)
+                return
+            }
+
             guard let error else {
                 completionHandler(nil)
                 return
             }
 
-            let errorDescription = (error as? LocalizedError)?.localizedDescription ?? String(describing: error)
+            let handler = {
+                let errorDescription = (error as? LocalizedError)?.localizedDescription ?? String(describing: error)
 
-            os_log("Tunnel startup error: %{public}@", type: .error, errorDescription)
-            self?.controllerErrorStore.lastErrorMessage = errorDescription
-            self?.connectionStatus = .disconnected
+                os_log("Tunnel startup error: %{public}@", type: .error, errorDescription)
+                self.controllerErrorStore.lastErrorMessage = errorDescription
+                self.connectionStatus = .disconnected
 
-            completionHandler(error)
+                completionHandler(error)
+            }
+
+            if startupOptions.startupMethod == .automaticOnDemand {
+                DispatchQueue.main.asyncAfter(deadline: startTime + DispatchTimeInterval.seconds(10), execute: handler)
+            } else {
+                handler()
+            }
```

Comment:
This is the gist of the changes above... we're executing a completion handler immediately or with a certain delay depending on the circumstances.